### PR TITLE
BUG: io/matlab: preserve dimensions of empty >=2D arrays

### LINKED ISSUE #13339: savemat discards dimension information if any dimension is zero
My issue is about `scipy.io.savemat`

#### Reproducing code example:
```
scipy.io.savemat(r'D:/temp/temp.mat', {'data': np.full((10, 0), np.nan)})
assert scipy.io.loadmat(r'D:/temp/temp.mat')['data'].shape == (10, 0)
```

#### Error message:
This fails because the loaded `data` variable has size `(0, 0)`

This is a problem with `savemat` but not `loadmat` because if I use Matlab to write out a `10x0` matrix it loads just fine:
```
scipy.io.loadmat(r'D:/temp/temp-from-matlab.mat')['data'].shape == (10, 0)
```

You can create this example file as follows:
```
$ cat >temp.inp <<'EOF'
TUFUTEFCIDUuMCBNQVQtZmlsZSwgUGxhdGZvcm06IFBDV0lONjQsIENyZWF0ZWQgb246IE1vbiBK
YW4gIDQgMDk6NDc6MzQgMjAyMSAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg
ICAAAAAAAAAAAAABSU0PAAAAJQAAAHic42NgYDAAYjYg5oDSIMAK5XNB+YwMLAwpiSWJTFA+ACQM
Ags=
EOF

$ base64 --decode temp.inp > temp-from-matlab.mat
```

A similar bug with empty arrays in `savemat` was fixed in #6096, maybe this is related.

#### Scipy/Numpy/Python version information:
```
1.5.2 1.19.2 sys.version_info(major=3, minor=7, micro=9, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by batterseapower**
#### Reference issue
Closes #13339

#### What does this implement/fix?
Saving 2D/3D/etc arrays to a Matlab file did not preserve the lengths of the dimensions, instead setting all the dimensions to zero length. This is inconsistent with Matlab itself.

### CODE REVIEW

File: scipy/io/matlab/tests/test_miobase.py [hunk 1]
```diff
@@ -11,18 +11,19 @@
 
 def test_matdims():
     # Test matdims dimension finder
-    assert_equal(matdims(np.array(1)), (1, 1))  # NumPy scalar
-    assert_equal(matdims(np.array([1])), (1, 1))  # 1-D array, 1 element
-    assert_equal(matdims(np.array([1,2])), (2, 1))  # 1-D array, 2 elements
-    assert_equal(matdims(np.array([[2],[3]])), (2, 1))  # 2-D array, column vector
-    assert_equal(matdims(np.array([[2,3]])), (1, 2))  # 2-D array, row vector
+    assert_equal(matdims(np.array(1)), (1, 1))  # numpy scalar
```
Comment (rgommers) [inline scipy/io/matlab/tests/test_miobase.py:14]: Can you undo the changes to comments in this file? The original spellings are the preferred ones, and these comments changes don't serve a function in this PR..

File: scipy/io/matlab/tests/test_miobase.py [hunk 2]
```diff
@@ -11,18 +11,19 @@
...
+    assert_equal(matdims(np.array([[2,3]])), (1, 2))  # 2d array, row vector
     # 3d array, rowish vector
     assert_equal(matdims(np.array([[[2,3]]])), (1, 1, 2))
-    assert_equal(matdims(np.array([])), (0, 0))  # empty 1-D array
-    assert_equal(matdims(np.array([[]])), (0, 0))  # empty 2-D array
-    assert_equal(matdims(np.array([[[]]])), (0, 0, 0))  # empty 3-D array
+    assert_equal(matdims(np.array([])), (0, 0))  # empty 1d array
+    assert_equal(matdims(np.array([[]])), (1, 0))  # empty 2d
+    assert_equal(matdims(np.array([[[]]])), (1, 1, 0))  # empty 3d
+    assert_equal(matdims(np.empty((1, 0, 1))), (1, 0, 1))  # empty 3d
```
Comment (rgommers) [inline scipy/io/matlab/tests/test_miobase.py:24]: Looks like this is the one line that needs to be kept.

File: scipy/io/matlab/miobase.py [hunk 3]
```diff
@@ -295,10 +295,10 @@ def matdims(arr, oned_as='column'):
     shape = arr.shape
     if shape == ():  # scalar
         return (1,1)
-    if functools.reduce(operator.mul, shape) == 0:  # zero elememts
-        return (0,) * np.max([arr.ndim, 2])
     if len(shape) == 1:  # 1D
-        if oned_as == 'column':
+        if functools.reduce(operator.mul, shape) == 0:  # zero elememts
```
Comment (rgommers) [inline scipy/io/matlab/miobase.py:299]: I realize this is an existing check, but it's weird and if `shape` has length 1 it's even more weird. Why not just `arr.size == 0` or `shape[0] == 0`?

File: scipy/io/matlab/miobase.py [hunk 4]
```diff
@@ -295,10 +295,10 @@ def matdims(arr, oned_as='column'):
...
+            return (0,0)
```
Comment (rgommers) [inline scipy/io/matlab/miobase.py:300]: minor: space after comma

**Review by rgommers — COMMENTED** [review]
Thanks @batterseapower. The fix looks good, I just have some minor comments.

**Comment by batterseapower** [comment]
Could this be merged now?

**Review by rgommers — APPROVED** [review]
Thanks for the ping @batterseapower. This LGTM now, merging.
